### PR TITLE
Updates to project management/settings

### PR DESCRIPTION
**Pull request by richATavail**
Mostly improvements to different settings UIs. Also added ability to import/export settings.

### CODE REVIEW

File: avail/src/main/kotlin/avail/anvil/settings/TemplateExpansionsManager.kt [hunk 1]
```diff
@@ -122,13 +128,19 @@ class RootTemplatesPanel constructor(
...
+
 	/**
 	 * The panel that contains all the shortcuts.
 	 */
 	private val templatesPanel = JPanel().apply {
 		layout = BoxLayout(this, BoxLayout.Y_AXIS)
 		root.templates.forEach {
-			add(TemplateRow(
+			add(
+				TemplateRow(
```
Comment (markATavail) [inline avail/src/main/kotlin/avail/anvil/settings/TemplateExpansionsManager.kt:143]: indent

File: avail/src/main/kotlin/avail/anvil/settings/TemplateExpansionsManager.kt [hunk 2]
```diff
@@ -158,45 +170,287 @@ class RootTemplatesPanel constructor(
...
+					fileFilter = FileNameExtensionFilter("*.json", "json")
+					addChoosableFileFilter(
+						object : FileFilter()
+						{
+							override fun getDescription(): String
+							{
+								return "Settings File (*.json)"
+							}
+
+							override fun accept(f: File?): Boolean
```
Comment (markATavail) [inline avail/src/main/kotlin/avail/anvil/settings/TemplateExpansionsManager.kt:319]: Ah, see my earlier comment.  Maybe revisit every place that matches `accept\(f: File\?`, and leave off the `?` for simplicity  It looks like there are 6 places.

File: avail/src/main/kotlin/avail/anvil/settings/TemplateSettings.kt [hunk 3]
```diff
@@ -0,0 +1,209 @@
+/*
+ * ShortcutSettings.kt
```
Comment (markATavail) [inline avail/src/main/kotlin/avail/anvil/settings/TemplateSettings.kt:2]: Update

File: avail/src/main/kotlin/avail/anvil/settings/TemplateSettings.kt [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+		 * The system's default [TemplateSettings] or `null` if there is a
+		 * problem retrieving the default [TemplateSettings].
+		 */
+		val systemDefaultTemplates: TemplateSettings?
+			get() =
+				try
+				{
+					readFromFile(File(
+						TemplateSettings::class.java.getResource(
+							"/defaultTemplates.json")!!.path))
```
Comment (markATavail) [inline avail/src/main/kotlin/avail/anvil/settings/TemplateSettings.kt:118]: Hoist into its own const val.

**Review by markATavail — CHANGES_REQUESTED** [review]

File: avail/src/main/kotlin/avail/anvil/AvailWorkbench.kt [hunk 5]
```diff
@@ -281,8 +282,20 @@ class AvailWorkbench internal constructor(
...
+	 */
+	fun saveProjectFileToDisk ()
+	{
+		val projectFilePath = availProjectFilePath
+		if (projectFilePath.isNotEmpty())
+		{
+			// Update the backing project file.
+			val writer = JSONWriter.newPrettyPrinterWriter()
+			availProject.writeTo(writer)
+			File(projectFilePath).writeText(writer.contents())
```
[the remark this answers is not in the document]
Comment (richATavail) [reply avail/src/main/kotlin/avail/anvil/AvailWorkbench.kt:296]: Done.

File: avail/src/main/kotlin/avail/anvil/settings/ShortcutManager.kt [hunk 6]
```diff
@@ -163,13 +170,145 @@ class ShortcutsPanel constructor(
...
+						this@ShortcutsPanel,
+						"Select Settings File")
+					if (result == JFileChooser.APPROVE_OPTION)
+					{
+						val shortcuts =
+							ShortcutSettings.readFromFile(selectedFile)
+						if (shortcuts != null)
+						{
+							workbench.globalAvailConfiguration
+								.attemptShortcutImport(shortcuts).let { m ->
```
[the remark this answers is not in the document]
Comment (richATavail) [reply avail/src/main/kotlin/avail/anvil/settings/ShortcutManager.kt:212]: I hoisted out the content of the `JFileChooser.apply` blocks to a private function:
```kotlin
private fun JFileChooser.importSettings ()
private fun JFileChooser.exportSettings ()
```
This provided a lot of tab relief.

File: avail/src/main/kotlin/avail/anvil/settings/TemplateExpansionsManager.kt (hunk 2 above)
Comment (richATavail) [reply avail/src/main/kotlin/avail/anvil/settings/TemplateExpansionsManager.kt:319]: I updated it everywhere.

**Comment by richATavail** [comment]
@markATavail I have finished the updates per your request. It is ready to merge.

**Pull request merged**